pytissueoptics/scene/intersection/intersectionFinder.py

```diff
@@ -53,25 +77,53 @@ def _findBBoxIntersectingSolids(self, ray) -> Union[List[Tuple[float, Solid]], N
-                closestIntersection = intersection
-                closestPolygon = polygon
-        if not closestIntersection:
-            return None
-        return Intersection(closestDistance, closestIntersection, closestPolygon)
 
-    def _findPolygonIntersection(self, ray: Ray, polygon: Polygon) -> Union[Vector, None]:
-        if isinstance(polygon, Triangle):
-            return self._triangleIntersect.getIntersection(ray, polygon)
-        if isinstance(polygon, Quad):
-            return self._quadIntersect.getIntersection(ray, polygon)
+class FastIntersectionFinder(IntersectionFinder):
+    def __init__(self, solids: List[Solid], partition: SpacePartition):
+        super(FastIntersectionFinder, self).__init__(solids)
+        self._partition = partition
+
+    def findIntersection(self, ray: Ray) -> Optional[Intersection]:
+        """
+        This is a simple home-made algorithm.
+
+        It is part of a recursive backtrack algorithm family. First, the ray origin is found with a recursive point search,
+        Then the intersection is found by checking is the ray intersects the neighbour node bbox. If it does, this node is explored.
+        When a leaf node is reached, the polygons within that leaf node are explored for a ray-polygon intersection.
+        If a hit is computed, the search is halted. Else, it continues to backtrack.
+
+        Limitations:    - does not take in consideration if the touched polygon is shared amongst many nodes
+        """
+
+        rayStartingNode = self._partition.searchPoint(ray.origin)
+        if rayStartingNode is None:
+            if self._boxIntersect.getIntersection(ray, self._partition.root.bbox):
+                rayStartingNode = self._partition.root
+            else:
+                return None
+        intersection = self._exploreNodeForIntersection(ray, rayStartingNode)
+        return intersection
+
+    def _exploreNodeForIntersection(self, ray: Ray, node: Node = None) -> Optional[Intersection]:
+        if not node.isLeaf:
+            closestIntersection = None
+            for child in node.children:
+                if not child.visited:
+                    bboxIntersection = self._boxIntersect.getIntersection(ray, child.bbox)
+                    child.visited = True
+                    if bboxIntersection is not None:
+                        if closestIntersection is not None:
+                            if (bboxIntersection - ray.origin).getNorm() > closestIntersection.distance:
+                                continue
+                        intersection = self._exploreNodeForIntersection(ray, child)
+                        if intersection is not None:
+                            if closestIntersection is None:
+                                closestIntersection = intersection
+                            elif intersection.distance < closestIntersection.distance:
+                                closestIntersection = intersection
```

Comment:
Not sure what this elif is for, but it is not covered by the tests.

Reply:
In the event that a new `bbox` is hit and that the hit was closer than the last closest surface, we must verify that the true intersection is truly closer. It might be the case that the `bbox` is hit, but the actual interface is further away. (Not our case actually because `bbox` cannot touch each other, but it is a necessary verication that would have been forgotten in the future.